File: Frontend/src/assets/fable_build/Analytics/Metrics.js
import { tryFindIndex, zip, setItem, fill, item, sumBy, average, pairwise, map } from "../fable_modules/fable-library-js.4.27.0/Array.js";
import { max } from "../fable_modules/fable-library-js.4.27.0/Double.js";
import { ofArray } from "../fable_modules/fable-library-js.4.27.0/Map.js";
import { comparePrimitives } from "../fable_modules/fable-library-js.4.27.0/Util.js";
import { SingleRunMetrics, DrawdownStats } from "./AnalyticsTypes.js";
import { defaultArg } from "../fable_modules/fable-library-js.4.27.0/Option.js";
import { sumBy as sumBy_1 } from "../fable_modules/fable-library-js.4.27.0/List.js";

const TRADING_DAYS = 252;

function calculateCAGR(startValue, endValue, days) {
    if (((startValue <= 0) ? true : (endValue <= 0)) ? true : (days <= 0)) {
        return 0;
    }
    else {
        const years = days / TRADING_DAYS;
        return Math.pow(endValue / startValue, 1 / years) - 1;
    }
}

function calculateLogReturns(curve) {
    return map((tupledArg) => {
        const prev = tupledArg[0];
        const curr = tupledArg[1];
        if ((prev <= 0) ? true : (curr <= 0)) {
            return 0;
        }
        else {
            return Math.log(curr / prev);
        }
    }, pairwise(curve), Float64Array);
}

function calculateVolatility(returns) {
    if (returns.length < 2) {
        return 0;
    }
    else {
        const mean = average(returns, {
            GetZero: () => 0,
            Add: (x_1, y) => (x_1 + y),
            DivideByInt: (x, i) => (x / i),
        });
        const sumSq = sumBy((r) => Math.pow(r - mean, 2), returns, {
            GetZero: () => 0,
            Add: (x_2, y_1) => (x_2 + y_1),
        });
        return Math.sqrt(sumSq / (returns.length - 1)) * Math.sqrt(TRADING_DAYS);
    }
}

function calculateDownsideDeviation(returns, targetReturnDaily) {
    if (returns.length < 2) {
        return 0;
    }
    else {
        const downsideSq = sumBy((r) => {
            const diff = r - targetReturnDaily;
            if (diff < 0) {
                return Math.pow(diff, 2);
            }
            else {
                return 0;
            }
        }, returns, {
            GetZero: () => 0,
            Add: (x, y) => (x + y),
        });
        return Math.sqrt(downsideSq / (returns.length - 1)) * Math.sqrt(TRADING_DAYS);
    }
}

function calculateDrawdownStats(curve) {
    let peak = item(0, curve);
    let maxDrawdown = 0;
    const thresholds = new Float64Array([0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1]);
    let counts = map((_arg) => 0, thresholds, Int32Array);
    let thresholdTriggered = map((_arg_1) => false, thresholds);
    for (let idx = 0; idx <= (curve.length - 1); idx++) {
        const value = item(idx, curve);
        if (value > peak) {
            peak = value;
            fill(thresholdTriggered, 0, thresholdTriggered.length, false);
        }
        else {
            const dd = (peak - value) / peak;
            maxDrawdown = max(maxDrawdown, dd);
            for (let i = 0; i <= (thresholds.length - 1); i++) {
                if ((dd >= item(i, thresholds)) && !item(i, thresholdTriggered)) {
                    setItem(counts, i, (item(i, counts) + 1) | 0);
                    setItem(thresholdTriggered, i, true);
                }
            }
        }
    }
    const countsMap = ofArray(zip(thresholds, counts), {
        Compare: comparePrimitives,
    });
    return new DrawdownStats(maxDrawdown, countsMap);
}

export function calculateSingleRun(run, config) {
    const curve = run.EquityCurve;
    const days = (curve.length - 1) | 0;
    const startVal = item(0, curve);
    const endVal = item(days, curve);
    const cagr = calculateCAGR(startVal, endVal, days);
    const logReturns = calculateLogReturns(curve);
    const vol = calculateVolatility(logReturns);
    const excessReturn = cagr - config.RiskFreeRate;
    const sharpe = (vol > 0) ? (excessReturn / vol) : 0;
    const downsideVol = calculateDownsideDeviation(logReturns, config.RiskFreeRate / TRADING_DAYS);
    const sortino = (downsideVol > 0) ? (excessReturn / downsideVol) : 0;
    const ddStats = calculateDrawdownStats(curve);
    const target = defaultArg(config.TargetWealth, 1.7976931348623157E+308);
    const daysToGoal = tryFindIndex((v) => (v >= target), curve);
    return new SingleRunMetrics(run.RunId, endVal, cagr, vol, sharpe, sortino, ddStats, daysToGoal != null, daysToGoal, curve.some((v_1) => (v_1 <= 0)), sumBy_1((t) => t.Commission, run.TransactionHistory, {
        GetZero: () => 0,
        Add: (x, y) => (x + y),
    }), sumBy_1((t_1) => t_1.Slippage, run.TransactionHistory, {
        GetZero: () => 0,
        Add: (x_1, y_1) => (x_1 + y_1),
    }));
}

File: Backend/StrategyEngine/Analytics/Metrics.fs
module StrategyEngine.Analytics.Metrics

open System
open StrategyEngine.Analytics.Types
open EngineTypes

let private TRADING_DAYS = 252.0

// --- Math Helpers ---

let private calculateCAGR (startValue: float) (endValue: float) (days: int) : float =
    if startValue <= 0.0 || endValue <= 0.0 || days <= 0 then 0.0
    else
        let years = float days / TRADING_DAYS
        Math.Pow(endValue / startValue, 1.0 / years) - 1.0

let private calculateLogReturns (curve: float[]) : float[] =
    curve
    |> Array.pairwise
    |> Array.map (fun (prev, curr) -> 
        if prev <= 0.0 || curr <= 0.0 then 0.0 
        else Math.Log(curr / prev))

let private calculateVolatility (returns: float[]) : float =
    if returns.Length < 2 then 0.0
    else
        let mean = Array.average returns
        let sumSq = returns |> Array.sumBy (fun r -> Math.Pow(r - mean, 2.0))
        let stdDev = Math.Sqrt(sumSq / float (returns.Length - 1))
        stdDev * Math.Sqrt(TRADING_DAYS)

let private calculateDownsideDeviation (returns: float[]) (targetReturnDaily: float) : float =
    if returns.Length < 2 then 0.0
    else
        let downsideSq = 
            returns 
            |> Array.sumBy (fun r -> 
                let diff = r - targetReturnDaily
                if diff < 0.0 then Math.Pow(diff, 2.0) else 0.0)
        
        let downsideStdDev = Math.Sqrt(downsideSq / float (returns.Length - 1))
        downsideStdDev * Math.Sqrt(TRADING_DAYS)

// --- Drawdown Logic ---

let private calculateDrawdownStats (curve: float[]) : DrawdownStats =
    let mutable peak = curve.[0]
    let mutable maxDrawdown = 0.0
    let thresholds = [| 0.1; 0.2; 0.3; 0.4; 0.5; 0.6; 0.7; 0.8; 0.9; 1.0 |]
    let mutable counts = thresholds |> Array.map (fun _ -> 0)
    let mutable thresholdTriggered = thresholds |> Array.map (fun _ -> false)

    for value in curve do
        if value > peak then
            peak <- value
            Array.fill thresholdTriggered 0 thresholdTriggered.Length false
        else
            let dd = (peak - value) / peak
            maxDrawdown <- Math.Max(maxDrawdown, dd)
            for i in 0 .. thresholds.Length - 1 do
                if dd >= thresholds.[i] && not thresholdTriggered.[i] then
                    counts.[i] <- counts.[i] + 1
                    thresholdTriggered.[i] <- true

    let countsMap = Array.zip thresholds counts |> Map.ofArray
    { MaxDrawdown = maxDrawdown; DrawdownCounts = countsMap }

// --- Main Calculator ---

let calculateSingleRun (run: SimulationRunResult) (config: AnalysisConfiguration) : SingleRunMetrics =
    let curve = run.EquityCurve
    let days = curve.Length - 1
    let startVal = curve.[0]
    let endVal = curve.[days]
    
    let cagr = calculateCAGR startVal endVal days
    let logReturns = calculateLogReturns curve
    let vol = calculateVolatility logReturns
    
    let excessReturn = cagr - config.RiskFreeRate
    let sharpe = if vol > 0.0 then excessReturn / vol else 0.0
    
    let dailyRf = config.RiskFreeRate / TRADING_DAYS
    let downsideVol = calculateDownsideDeviation logReturns dailyRf
    let sortino = if downsideVol > 0.0 then excessReturn / downsideVol else 0.0

    let ddStats = calculateDrawdownStats curve

    let target = Option.defaultValue Double.MaxValue config.TargetWealth
    let daysToGoal = curve |> Array.tryFindIndex (fun v -> v >= target)
    let reachedGoal = daysToGoal.IsSome
    let isRuined = curve |> Array.exists (fun v -> v <= 0.0)

    // Sum Costs
    let totalComm = run.TransactionHistory |> List.sumBy (fun t -> t.Commission)
    let totalSlip = run.TransactionHistory |> List.sumBy (fun t -> t.Slippage)
    let totalTax = run.TransactionHistory |> List.sumBy (fun t -> t.Tax) // <--- Added

    {
        RunId = run.RunId
        FinalWealth = endVal
        CAGR = cagr
        AnnualizedVolatility = vol
        SharpeRatio = sharpe
        SortinoRatio = sortino
        Drawdown = ddStats
        ReachedGoal = reachedGoal
        DaysToGoal = daysToGoal
        IsRuined = isRuined
        TotalCommission = totalComm
        TotalSlippage = totalSlip
        TotalTax = totalTax // <--- Added
    }